content/en/blog/_posts/2025-02-14-cloud-controller-manager-chicken-egg-problem.md

```diff
@@ -0,0 +1,250 @@
+---
+layout: blog
+title: "The Cloud Controller Manager Chicken and Egg Problem"
+date: 2025-02-14
+slug: cloud-controller-manager-chicken-egg-problem
+author: >
+  Antonio Ojea,
+  Michael McCune
+draft: true
+---
+
+Kubernetes 1.31
+[completed the largest migration in Kubernetes history][migration-blog], removing the in-tree
+cloud provider.  While the component migration is now done, this leaves some additional
+complexity for users and installer projects (for example, kOps or Cluster API) .  We will go
+over those additional steps and failure points and make recommendations for cluster owners.
+This migration was complex and some logic had to be extracted from the core components,
+building four new subsystems.
+
+1. **Cloud controller manager** ([KEP-2392][kep2392])
+2. **API server network proxy** ([KEP-1281][kep1281])
+3. **kubelet credential provider plugins** ([KEP-2133][kep2133])
+4. **Storage migration to use [CSI][csi]** ([KEP-625][kep625])
+
+The [cloud controller manager is part of the control plane][ccm]. It is a critical component
+that replaces some functionality that existed previously in the kube-controller-manager and the
+kubelet.
+
+![Components of Kubernetes](https://kubernetes.io/images/docs/components-of-kubernetes.svg)
+
+One of the most critical functionalities of the cloud controller manager is the node controller,
+which is responsible for the initialization of the nodes.
+
+As you can see in the following diagram, when the **kubelet** starts, it registers the `Node`
+object with the apiserver, Tainting the node so it can be processed first by the
+cloud-controller-manager. The initial `Node` is missing the cloud-provider specific information,
+like the Node Addresses and the Labels with the cloud provider specific information like the
+Node, Region and Instance type information.
+
+```mermaid
+sequenceDiagram
+    autonumber
+    rect rgb(191, 223, 255)
+    Kubelet->>+Kube-apiserver: Create Node
+    Note over Kubelet: Taint:<br/> node.cloudprovider.kubernetes.io
+    Kube-apiserver->>-Kubelet: Node  Created
+    end
+    Note over Kube-apiserver: Node is Not Ready<br/> Tainted, Missing Node Addresses*, ...
+    Note over Kube-apiserver: Send Updates
+    rect rgb(200, 150, 255)
+    Kube-apiserver->>+Cloud-controller-manager: Watch: New Node Created
+    Note over Cloud-controller-manager: Initialize Node:<br/>Cloud Provider Labels, Node Addresses, ...
+    Cloud-controller-manager->>-Kube-apiserver: Update Node
+    end
+    Note over Kube-apiserver: Node is Ready
+```
```

Comment:
For maintainability: ideally, capture this to SVG and turn it into a static SVG + `figure` shortcode.

Reply:
i'm not familiar with the `figure` shortcode, do you have a docs link about it?

Reply:
See https://github.com/kubernetes/website/pull/49612#issuecomment-2659827569 and https://kubernetes.io/docs/contribute/style/diagram-guide/